File: C++/1039-Course_list_for_student.cc
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>


int main() {

    std::ios::sync_with_stdio(false);

    int n_query, n_course;
    std::cin >> n_query >> n_course;

    std::map<std::string, std::vector<int>> stu_to_course;
    while (n_course--) {
        int course_idx, n_stu;
        std::cin >> course_idx >> n_stu;
        while (n_stu--) {
            std::string stu;
            std::cin >> stu;
            stu_to_course[stu].push_back(course_idx);
        }
    }

    while (n_query--) {
        std::string stu;
        std::cin >> stu;
        std::vector<int> reg_courses = stu_to_course[stu];
        std::sort(reg_courses.begin(), reg_courses.end());
        std::cout << stu  << ' ' << reg_courses.size();
        for (int course_idx : reg_courses) {
            std::cout << ' ' << course_idx;
        }
        std::cout << '\n';
    }

    return 0;
}

File: Python/1039-Course_list_for_student.py
from collections import defaultdict


if __name__ == '__main__':
    _, n_course = [int(i) for i in input().split()]
    stu_to_course = defaultdict(list)
    for _ in range(n_course):
        course_idx, n_stu = [int(i) for i in input().split()]
        if n_stu:
            for stu in input().split():
                stu_to_course[stu].append(course_idx)
    for query in input().split():
        print(query,
              len(stu_to_course[query]),
              *sorted(stu_to_course[query]))
